pytest_pdb.py

```diff
@@ -3,31 +3,80 @@
 import pytest
 
 
-def find_test(currentframe):
-    frame = currentframe
-    prev = frame
-    while frame:
-        for value in frame.f_locals.values():
-            if isinstance(value, pytest.Item):
-                return (value, prev)
-        prev = frame
-        frame = frame.f_back
+def find_test(stack):
+    c = 0
+    for st in stack:
+        frame = st[0]
+        c += 1
+        if frame.f_code.co_name == 'pytest_pyfunc_call':
+            if 'pyfuncitem' in frame.f_locals:
+                item = frame.f_locals['pyfuncitem']
+                if isinstance(item, pytest.Item):
+                    return (item, stack[c][0], c)
 
 
 class PdbExtension:
     def do_whichtest(self, arg):
         """whichtest
         Show which test we are currently in.
         """
-        test = find_test(self.curframe)
+        test = find_test(self.stack)
         if test is None:
-            print("Couldn't determine current test", file=self.stdout)
+            print("Couldn't determine current test.", file=self.stdout)
             return
+        (item, frame, stack_idx) = test
 
-        (item, frame) = test
-        print("Currently in {} ({}:{}) on line {}".format(
+
+        # Describe offset to current frame (for up/down).
+        cur_idx = None
+        for i, st in enumerate(self.stack, start=1):
+            if st[0] == self.curframe:
+                cur_idx = i
+                break
+        if cur_idx is None:
+            print("warning: couldn't determine current frame.",
+                  file=self.stdout)
+        else:
+            offset = stack_idx + 1 - cur_idx
+            if offset == 0:
+                desc = ' (at current frame)'
+            elif offset > 0:
+                desc = ' ({} frames down)'.format(offset)
+            else:
+                desc = ' ({} frames up)'.format(0 - offset)
+
+        print("Currently in {} ({}:{}) on line {}{}.".format(
             item.location[2], item.location[0], item.location[1] + 1,
-            frame.f_lineno), file=self.stdout)
+            frame.f_lineno, desc), file=self.stdout)
+
+    def do_gototest(self, arg):
+        """gototest
+        Go to frame containing the test.
+        """
+        test_caller = find_test(self.stack)
+        if test_caller is None:
+            print("Couldn't determine current test.", file=self.stdout)
+            return
+
+        self._select_frame(test_caller[2])
+
+    def do_top(self, arg):
```

Comment:
What exactly do top and bottom relate to?

Is top the test, or actually the first frame in the whole stack? Not sure I understand the use case for it.

For bottom I can see the use. I often go up several frames, then do 'd' until I get a message that there is no more frame.

Reply:
`top` is the first frame in the whole stack.
Inspired by https://stackoverflow.com/questions/38619411/go-to-top-frame-in-pdb-python-debugger.